File: src/game/capture.ts
import type { Capture, Player, Point, Stone } from "./types";

interface Face {
  area: number;
  boundary: Point[];
  key: string;
}

const OFFSETS: Point[] = [
  { x: -1, y: -1 },
  { x: 0, y: -1 },
  { x: 1, y: -1 },
  { x: -1, y: 0 },
  { x: 1, y: 0 },
  { x: -1, y: 1 },
  { x: 0, y: 1 },
  { x: 1, y: 1 }
];

const pointKey = ({ x, y }: Point): string => `${x}:${y}`;
const samePoint = (a: Point, b: Point): boolean => a.x === b.x && a.y === b.y;
const edgeKey = (a: Point, b: Point): string => `${pointKey(a)}>${pointKey(b)}`;

const mathAngle = (from: Point, to: Point): number => Math.atan2(-(to.y - from.y), to.x - from.x);

const clockwiseDelta = (from: number, to: number): number => {
  const full = Math.PI * 2;
  const delta = ((from - to) % full + full) % full;
  return delta < 1e-9 ? full : delta;
};

const signedMathArea = (polygon: Point[]): number => {
  let doubled = 0;
  for (let index = 0; index < polygon.length; index += 1) {
    const current = polygon[index];
    const next = polygon[(index + 1) % polygon.length];
    doubled += current.x * -next.y - next.x * -current.y;
  }
  return doubled / 2;
};

const cross = (a: Point, b: Point, c: Point): number =>
  (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x);

const onSegment = (a: Point, b: Point, point: Point): boolean =>
  Math.min(a.x, b.x) <= point.x &&
  point.x <= Math.max(a.x, b.x) &&
  Math.min(a.y, b.y) <= point.y &&
  point.y <= Math.max(a.y, b.y);

const pointOnPolygonBoundary = (point: Point, polygon: Point[]): boolean => {
  for (let index = 0; index < polygon.length; index += 1) {
    const current = polygon[index];
    const next = polygon[(index + 1) % polygon.length];
    if (cross(current, next, point) === 0 && onSegment(current, next, point)) return true;
  }
  return false;
};

const segmentsIntersect = (a: Point, b: Point, c: Point, d: Point): boolean => {
  const abC = cross(a, b, c);
  const abD = cross(a, b, d);
  const cdA = cross(c, d, a);
  const cdB = cross(c, d, b);

  if (abC === 0 && onSegment(a, b, c)) return true;
  if (abD === 0 && onSegment(a, b, d)) return true;
  if (cdA === 0 && onSegment(c, d, a)) return true;
  if (cdB === 0 && onSegment(c, d, b)) return true;

  return (abC > 0) !== (abD > 0) && (cdA > 0) !== (cdB > 0);
};

const isSimplePolygon = (polygon: Point[]): boolean => {
  const keys = new Set(polygon.map(pointKey));
  if (keys.size !== polygon.length) return false;

  for (let first = 0; first < polygon.length; first += 1) {
    const firstNext = (first + 1) % polygon.length;
    for (let second = first + 1; second < polygon.length; second += 1) {
      const secondNext = (second + 1) % polygon.length;
      if (first === second || firstNext === second || secondNext === first) continue;
      if (first === 0 && secondNext === 0) continue;
      if (segmentsIntersect(polygon[first], polygon[firstNext], polygon[second], polygon[secondNext])) return false;
    }
  }

  return true;
};

const canonicalBoundaryKey = (boundary: Point[]): string => {
  const variants: string[] = [];
  const encode = (points: Point[]): string => points.map(pointKey).join("|");

  for (const points of [boundary, [...boundary].reverse()]) {
    for (let offset = 0; offset < points.length; offset += 1) {
      variants.push(encode([...points.slice(offset), ...points.slice(0, offset)]));
    }
  }

  variants.sort();
  return variants[0];
};

const capturedKeys = (captures: Capture[]): Set<string> =>
  new Set(captures.flatMap((capture) => capture.captured.map(pointKey)));

export const pointInPolygon = (point: Point, polygon: Point[]): boolean => {
  if (pointOnPolygonBoundary(point, polygon)) return false;

  let inside = false;
  for (let current = 0, previous = polygon.length - 1; current < polygon.length; previous = current, current += 1) {
    const a = polygon[current];
    const b = polygon[previous];
    const crosses =
      (a.y > point.y) !== (b.y > point.y) &&
      point.x < ((b.x - a.x) * (point.y - a.y)) / (b.y - a.y) + a.x;
    if (crosses) inside = !inside;
  }
  return inside;
};

const pointInOrOnPolygon = (point: Point, polygon: Point[]): boolean =>
  pointOnPolygonBoundary(point, polygon) || pointInPolygon(point, polygon);

export const pointInsideCapture = (point: Point, capture: Capture): boolean =>
  pointInPolygon(point, capture.boundary);

export const pointInsideAnyCapture = (point: Point, captures: Capture[]): boolean =>
  captures.some((capture) => pointInsideCapture(point, capture));

const nextNeighbor = (previous: Point, current: Point, neighbors: Point[]): Point | undefined => {
  const reverseAngle = mathAngle(current, previous);
  let best: Point | undefined;
  let bestDelta = Number.POSITIVE_INFINITY;

  for (const candidate of neighbors) {
    const delta = clockwiseDelta(reverseAngle, mathAngle(current, candidate));
    if (delta < bestDelta) {
      best = candidate;
      bestDelta = delta;
    }
  }

  return best;
};

const extractFaces = (stones: Map<string, Stone>, owner: Player, excluded: Set<string>): Face[] => {
  const ownerStones = [...stones.values()].filter(
    (stone) => stone.player === owner && !excluded.has(pointKey(stone))
  );
  const byKey = new Map(ownerStones.map((stone) => [pointKey(stone), stone]));
  const neighbors = new Map<string, Point[]>();

  for (const stone of ownerStones) {
    const adjacent: Point[] = [];
    for (const offset of OFFSETS) {
      const candidate = byKey.get(`${stone.x + offset.x}:${stone.y + offset.y}`);
      if (candidate) adjacent.push(candidate);
    }
    neighbors.set(pointKey(stone), adjacent);
  }

  const directedEdgeCount = [...neighbors.values()].reduce((sum, adjacent) => sum + adjacent.length, 0);
  const visited = new Set<string>();
  const faces = new Map<string, Face>();

  for (const start of ownerStones) {
    for (const first of neighbors.get(pointKey(start)) ?? []) {
      const startEdge = edgeKey(start, first);
      if (visited.has(startEdge)) continue;

      let previous: Point = start;
      let current: Point = first;
      const localEdges = new Set<string>();
      const boundary: Point[] = [];
      let closed = false;

      for (let step = 0; step <= directedEdgeCount + 1; step += 1) {
        const currentEdge = edgeKey(previous, current);
        if (localEdges.has(currentEdge)) {
          closed = currentEdge === startEdge;
          break;
        }

        localEdges.add(currentEdge);
        visited.add(currentEdge);
        boundary.push(previous);

        const following = nextNeighbor(previous, current, neighbors.get(pointKey(current)) ?? []);
        if (!following) break;

        previous = current;
        current = following;
        if (edgeKey(previous, current) === startEdge) {
          closed = true;
          break;
        }
      }

      if (!closed || boundary.length < 3 || !isSimplePolygon(boundary)) continue;

      const area = signedMathArea(boundary);
      if (area <= 0) continue;

      const key = canonicalBoundaryKey(boundary);
      faces.set(key, { area, boundary, key });
    }
  }

  return [...faces.values()].sort((a, b) => a.area - b.area || a.key.localeCompare(b.key));
};

const captureForFace = (
  face: Face,
  stones: Map<string, Stone>,
  owner: Player,
  excluded: Set<string>
): Capture | undefined => {
  const captured = [...stones.values()].filter(
    (stone) => stone.player !== owner && !excluded.has(pointKey(stone)) && pointInPolygon(stone, face.boundary)
  );
  return captured.length > 0 ? { owner, boundary: face.boundary, captured } : undefined;
};

export const findNewCaptures = (
  stones: Map<string, Stone>,
  existingCaptures: Capture[],
  owner: Player,
  closingPoint: Point
): Capture[] => {
  const excluded = capturedKeys(existingCaptures);
  const faces = extractFaces(stones, owner, excluded).filter((face) =>
    face.boundary.some((point) => samePoint(point, closingPoint))
  );
  const groups = new Map<string, Capture>();

  for (const stone of stones.values()) {
    if (stone.player === owner || excluded.has(pointKey(stone))) continue;

    const face = faces.find((candidate) => pointInPolygon(stone, candidate.boundary));
    if (!face) continue;

    const capture = groups.get(face.key) ?? { owner, boundary: face.boundary, captured: [] };
    capture.captured.push(stone);
    groups.set(face.key, capture);
  }

  return [...groups.values()];
};

export const findHouseCapture = (
  stones: Map<string, Stone>,
  existingCaptures: Capture[],
  owner: Player,
  intruder: Stone
): Capture | undefined => {
  if (intruder.player === owner) return undefined;

  const excluded = capturedKeys(existingCaptures);
  if (excluded.has(pointKey(intruder))) return undefined;

  const face = extractFaces(stones, owner, excluded).find((candidate) => pointInPolygon(intruder, candidate.boundary));
  return face ? captureForFace(face, stones, owner, excluded) : undefined;
};

const captureContainsCapture = (outer: Capture, inner: Capture): boolean =>
  inner.boundary.every((point) => pointInOrOnPolygon(point, outer.boundary));

const sameCaptureBoundary = (a: Capture, b: Capture): boolean =>
  a.owner === b.owner && canonicalBoundaryKey(a.boundary) === canonicalBoundaryKey(b.boundary);

export const applyCaptures = (existingCaptures: Capture[], newCaptures: Capture[]): Capture[] => {
  let active = [...existingCaptures];

  for (const capture of newCaptures) {
    active = active.filter(
      (existing) =>
        !sameCaptureBoundary(existing, capture) &&
        (existing.owner === capture.owner || !captureContainsCapture(capture, existing))
    );
    active.push(capture);
  }

  return active;
};

export const resolveCapturesAfterMove = (
  stones: Map<string, Stone>,
  existingCaptures: Capture[],
  player: Player,
  closingPoint: Point
): Capture[] => {
  const direct = findNewCaptures(stones, existingCaptures, player, closingPoint);
  if (direct.length > 0) return applyCaptures(existingCaptures, direct);

  const opponent: Player = player === "red" ? "blue" : "red";
  const intruder = stones.get(pointKey(closingPoint));
  if (!intruder) return existingCaptures;

  const house = findHouseCapture(stones, existingCaptures, opponent, intruder);
  return house ? applyCaptures(existingCaptures, [house]) : existingCaptures;
};

export const scoreCaptures = (captures: Capture[]): Record<Player, number> => {
  const red = new Set<string>();
  const blue = new Set<string>();

  for (const capture of captures) {
    const target = capture.owner === "red" ? red : blue;
    for (const stone of capture.captured) target.add(pointKey(stone));
  }

  return { red: red.size, blue: blue.size };
};

File: crates/game-core/src/capture.rs
use crate::types::{Capture, GameState, Player, Point, Score, Stone};
use std::collections::{HashMap, HashSet};
use std::f64::consts::PI;

#[derive(Clone)]
struct Face {
    area: f64,
    boundary: Vec<Point>,
    key: String,
}

const OFFSETS: [(i64, i64); 8] = [
    (-1, -1),
    (0, -1),
    (1, -1),
    (-1, 0),
    (1, 0),
    (-1, 1),
    (0, 1),
    (1, 1),
];

fn point_key(point: Point) -> String {
    format!("{}:{}", point.x, point.y)
}

fn edge_key(a: Point, b: Point) -> String {
    format!("{}>{}", point_key(a), point_key(b))
}

fn math_angle(from: Point, to: Point) -> f64 {
    (-(to.y - from.y) as f64).atan2((to.x - from.x) as f64)
}

fn clockwise_delta(from: f64, to: f64) -> f64 {
    let full = PI * 2.0;
    let delta = (from - to).rem_euclid(full);
    if delta < 1e-9 {
        full
    } else {
        delta
    }
}

fn signed_math_area(polygon: &[Point]) -> f64 {
    let mut doubled = 0.0;
    for index in 0..polygon.len() {
        let current = polygon[index];
        let next = polygon[(index + 1) % polygon.len()];
        doubled += current.x as f64 * -(next.y as f64) - next.x as f64 * -(current.y as f64);
    }
    doubled / 2.0
}

fn cross(a: Point, b: Point, c: Point) -> i128 {
    (b.x - a.x) as i128 * (c.y - a.y) as i128 - (b.y - a.y) as i128 * (c.x - a.x) as i128
}

fn on_segment(a: Point, b: Point, point: Point) -> bool {
    a.x.min(b.x) <= point.x
        && point.x <= a.x.max(b.x)
        && a.y.min(b.y) <= point.y
        && point.y <= a.y.max(b.y)
}

fn point_on_polygon_boundary(point: Point, polygon: &[Point]) -> bool {
    if polygon.is_empty() {
        return false;
    }
    for index in 0..polygon.len() {
        let current = polygon[index];
        let next = polygon[(index + 1) % polygon.len()];
        if cross(current, next, point) == 0 && on_segment(current, next, point) {
            return true;
        }
    }
    false
}

fn segments_intersect(a: Point, b: Point, c: Point, d: Point) -> bool {
    let ab_c = cross(a, b, c);
    let ab_d = cross(a, b, d);
    let cd_a = cross(c, d, a);
    let cd_b = cross(c, d, b);

    if ab_c == 0 && on_segment(a, b, c) {
        return true;
    }
    if ab_d == 0 && on_segment(a, b, d) {
        return true;
    }
    if cd_a == 0 && on_segment(c, d, a) {
        return true;
    }
    if cd_b == 0 && on_segment(c, d, b) {
        return true;
    }

    (ab_c > 0) != (ab_d > 0) && (cd_a > 0) != (cd_b > 0)
}

fn is_simple_polygon(polygon: &[Point]) -> bool {
    let keys: HashSet<Point> = polygon.iter().copied().collect();
    if keys.len() != polygon.len() {
        return false;
    }

    for first in 0..polygon.len() {
        let first_next = (first + 1) % polygon.len();
        for second in first + 1..polygon.len() {
            let second_next = (second + 1) % polygon.len();
            if first == second || first_next == second || second_next == first {
                continue;
            }
            if first == 0 && second_next == 0 {
                continue;
            }
            if segments_intersect(
                polygon[first],
                polygon[first_next],
                polygon[second],
                polygon[second_next],
            ) {
                return false;
            }
        }
    }

    true
}

fn canonical_boundary_key(boundary: &[Point]) -> String {
    fn encode(points: &[Point]) -> String {
        points
            .iter()
            .map(|point| point_key(*point))
            .collect::<Vec<_>>()
            .join("|")
    }

    let mut variants = Vec::with_capacity(boundary.len() * 2);
    for reverse in [false, true] {
        let mut points = boundary.to_vec();
        if reverse {
            points.reverse();
        }
        for offset in 0..points.len() {
            let mut rotated = points[offset..].to_vec();
            rotated.extend_from_slice(&points[..offset]);
            variants.push(encode(&rotated));
        }
    }
    variants.sort();
    variants.into_iter().next().unwrap_or_default()
}

fn captured_keys(captures: &[Capture]) -> HashSet<Point> {
    captures
        .iter()
        .flat_map(|capture| capture.captured.iter().map(|stone| stone.point()))
        .collect()
}

pub fn point_in_polygon(point: Point, polygon: &[Point]) -> bool {
    if polygon.is_empty() || point_on_polygon_boundary(point, polygon) {
        return false;
    }

    let mut inside = false;
    let mut previous = polygon.len() - 1;
    for current in 0..polygon.len() {
        let a = polygon[current];
        let b = polygon[previous];
        let crosses = (a.y > point.y) != (b.y > point.y)
            && point.x as f64
                < (b.x - a.x) as f64 * (point.y - a.y) as f64 / (b.y - a.y) as f64
                    + a.x as f64;
        if crosses {
            inside = !inside;
        }
        previous = current;
    }
    inside
}

fn point_in_or_on_polygon(point: Point, polygon: &[Point]) -> bool {
    point_on_polygon_boundary(point, polygon) || point_in_polygon(point, polygon)
}

pub fn point_inside_capture(point: Point, capture: &Capture) -> bool {
    point_in_polygon(point, &capture.boundary)
}

pub fn point_inside_any_capture(point: Point, captures: &[Capture]) -> bool {
    captures.iter().any(|capture| point_inside_capture(point, capture))
}

fn next_neighbor(previous: Point, current: Point, neighbors: &[Point]) -> Option<Point> {
    let reverse_angle = math_angle(current, previous);
    let mut best = None;
    let mut best_delta = f64::INFINITY;

    for candidate in neighbors {
        let delta = clockwise_delta(reverse_angle, math_angle(current, *candidate));
        if delta < best_delta {
            best = Some(*candidate);
            best_delta = delta;
        }
    }

    best
}

fn extract_faces(state: &GameState, owner: Player, excluded: &HashSet<Point>) -> Vec<Face> {
    let owner_stones: Vec<Stone> = state
        .stones
        .iter()
        .copied()
        .filter(|stone| stone.player == owner && !excluded.contains(&stone.point()))
        .collect();
    let by_point: HashMap<Point, Stone> = owner_stones.iter().copied().map(|stone| (stone.point(), stone)).collect();
    let mut neighbors: HashMap<Point, Vec<Point>> = HashMap::new();

    for stone in &owner_stones {
        let mut adjacent = Vec::new();
        for (dx, dy) in OFFSETS {
            let point = Point {
                x: stone.x + dx,
                y: stone.y + dy,
            };
            if by_point.contains_key(&point) {
                adjacent.push(point);
            }
        }
        neighbors.insert(stone.point(), adjacent);
    }

    let directed_edge_count: usize = neighbors.values().map(Vec::len).sum();
    let mut visited: HashSet<String> = HashSet::new();
    let mut faces: HashMap<String, Face> = HashMap::new();

    for start in &owner_stones {
        let start_point = start.point();
        for first in neighbors.get(&start_point).into_iter().flatten() {
            let start_edge = edge_key(start_point, *first);
            if visited.contains(&start_edge) {
                continue;
            }

            let mut previous = start_point;
            let mut current = *first;
            let mut local_edges = HashSet::new();
            let mut boundary = Vec::new();
            let mut closed = false;

            for _ in 0..=directed_edge_count + 1 {
                let current_edge = edge_key(previous, current);
                if local_edges.contains(&current_edge) {
                    closed = current_edge == start_edge;
                    break;
                }

                local_edges.insert(current_edge.clone());
                visited.insert(current_edge);
                boundary.push(previous);

                let following = neighbors
                    .get(&current)
                    .and_then(|items| next_neighbor(previous, current, items));
                let Some(following) = following else {
                    break;
                };

                previous = current;
                current = following;
                if edge_key(previous, current) == start_edge {
                    closed = true;
                    break;
                }
            }

            if !closed || boundary.len() < 3 || !is_simple_polygon(&boundary) {
                continue;
            }

            let area = signed_math_area(&boundary);
            if area <= 0.0 {
                continue;
            }

            let key = canonical_boundary_key(&boundary);
            faces.insert(
                key.clone(),
                Face {
                    area,
                    boundary,
                    key,
                },
            );
        }
    }

    let mut result: Vec<Face> = faces.into_values().collect();
    result.sort_by(|a, b| {
        a.area
            .partial_cmp(&b.area)
            .unwrap_or(std::cmp::Ordering::Equal)
            .then_with(|| a.key.cmp(&b.key))
    });
    result
}

fn capture_for_face(
    face: &Face,
    state: &GameState,
    owner: Player,
    excluded: &HashSet<Point>,
) -> Option<Capture> {
    let captured: Vec<Stone> = state
        .stones
        .iter()
        .copied()
        .filter(|stone| {
            stone.player != owner
                && !excluded.contains(&stone.point())
                && point_in_polygon(stone.point(), &face.boundary)
        })
        .collect();

    (!captured.is_empty()).then(|| Capture {
        owner,
        boundary: face.boundary.clone(),
        captured,
    })
}

pub fn find_new_captures(
    state: &GameState,
    existing_captures: &[Capture],
    owner: Player,
    closing_point: Point,
) -> Vec<Capture> {
    let excluded = captured_keys(existing_captures);
    let faces: Vec<Face> = extract_faces(state, owner, &excluded)
        .into_iter()
        .filter(|face| face.boundary.contains(&closing_point))
        .collect();
    let mut groups: Vec<(String, Capture)> = Vec::new();

    for stone in &state.stones {
        if stone.player == owner || excluded.contains(&stone.point()) {
            continue;
        }
        let Some(face) = faces
            .iter()
            .find(|candidate| point_in_polygon(stone.point(), &candidate.boundary))
        else {
            continue;
        };

        if let Some((_, capture)) = groups.iter_mut().find(|(key, _)| *key == face.key) {
            capture.captured.push(*stone);
        } else {
            groups.push((
                face.key.clone(),
                Capture {
                    owner,
                    boundary: face.boundary.clone(),
                    captured: vec![*stone],
                },
            ));
        }
    }

    groups.into_iter().map(|(_, capture)| capture).collect()
}

pub fn find_house_capture(
    state: &GameState,
    existing_captures: &[Capture],
    owner: Player,
    intruder: Stone,
) -> Option<Capture> {
    if intruder.player == owner {
        return None;
    }

    let excluded = captured_keys(existing_captures);
    if excluded.contains(&intruder.point()) {
        return None;
    }

    extract_faces(state, owner, &excluded)
        .iter()
        .find(|candidate| point_in_polygon(intruder.point(), &candidate.boundary))
        .and_then(|face| capture_for_face(face, state, owner, &excluded))
}

fn capture_contains_capture(outer: &Capture, inner: &Capture) -> bool {
    inner
        .boundary
        .iter()
        .all(|point| point_in_or_on_polygon(*point, &outer.boundary))
}

fn same_capture_boundary(a: &Capture, b: &Capture) -> bool {
    a.owner == b.owner && canonical_boundary_key(&a.boundary) == canonical_boundary_key(&b.boundary)
}

pub fn apply_captures(existing_captures: &[Capture], new_captures: &[Capture]) -> Vec<Capture> {
    let mut active = existing_captures.to_vec();

    for capture in new_captures {
        active.retain(|existing| {
            !same_capture_boundary(existing, capture)
                && (existing.owner == capture.owner || !capture_contains_capture(capture, existing))
        });
        active.push(capture.clone());
    }

    active
}

pub fn resolve_captures_after_move(
    state: &GameState,
    existing_captures: &[Capture],
    player: Player,
    closing_point: Point,
) -> Vec<Capture> {
    let direct = find_new_captures(state, existing_captures, player, closing_point);
    if !direct.is_empty() {
        return apply_captures(existing_captures, &direct);
    }

    let opponent = player.other();
    let Some(intruder) = state.stone_at(closing_point).copied() else {
        return existing_captures.to_vec();
    };

    match find_house_capture(state, existing_captures, opponent, intruder) {
        Some(house) => apply_captures(existing_captures, &[house]),
        None => existing_captures.to_vec(),
    }
}

pub fn score_captures(captures: &[Capture]) -> Score {
    let mut red = HashSet::new();
    let mut blue = HashSet::new();

    for capture in captures {
        let target = if capture.owner == Player::Red {
            &mut red
        } else {
            &mut blue
        };
        for stone in &capture.captured {
            target.insert(stone.point());
        }
    }

    Score {
        red: red.len(),
        blue: blue.len(),
    }
}
